### PR TITLE
fix torch.is_tensor not recognizing HalfTensor

### LINKED ISSUE #1922: torch.is_tensor(torch.HalfTensor()) returns False.
The problem is [here](https://github.com/pytorch/pytorch/blob/master/torch/__init__.py#L274).

### PR DESCRIPTION
**Pull request by ethanluoyc**
Fix #1922.

### CODE REVIEW

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
This needs some additional changes in the tests

**Comment by ethanluoyc** [comment]
@apaszke Hmm now I am bit curious about the reasons for the failure of the test cases. It says that ne and fill are not implemented for the HalfTensor. From from the source code I think we are actually generating those methods since there isn't any restriction in the cwrap definitions?

**Comment by soumith** [comment]
@ethanluoyc CPU HalfTensor has no methods implemented except copy and serialization stuff. So almost all tests have to be filtered out for CPU HalfTensor

**Comment by gchanan** [comment]
what we do in TH is basically define an "is_tensor" and a "has_tensor_math"; then you write the tests based on which of those you are actually testing.

**Comment by ethanluoyc** [comment]
@soumith so I took a look at `test_torch.py`, it looks like the only reference to tensor_classes is in test_print. For other tests usually they are self-contained about which types they are testing against.

**Comment by ethanluoyc** [comment]
After fixing this issue. I wonder why it would be a good idea to have a CPU HalfTensor at all. Since HalfTensor is a subtype of Tensor, we would expect it to support the same interface as Tensor. However, the current implementation of HalfTensor does not allow such behaviour.

**Comment by soumith** [comment]
on the CPU, HalfTensor is just there for serialization purposes. For example saving a GPU HalfTensor and being able to load the weights back on a CPU-only machine.

Other than that it has no purpose, as Half is not a native type on CPU and all operations will be very very very slow.

File: test/test_torch.py [hunk 1]
```diff
@@ -3590,20 +3590,20 @@ def test_half_tensor(self):
...
+        self.assertEqual(x.cuda(), x)
 
         xc = x.cuda()
         with tempfile.NamedTemporaryFile() as f:
             torch.save(xc, f)
             f.seek(0)
             xc2 = torch.load(f)
             self.assertIsInstance(xc2, type(xc))
-            self.assertEqual(xc, xc2)
+            self.assertEqual(xc.float(), xc2.float())
```
Comment (apaszke) [inline test/test_torch.py]: Why was it necessary to change them to float tensors? assertEqual should work for HalfTensors

Comment (ethanluoyc) [inline test/test_torch.py]: Hmm @apaszke it looks like HalfTensor on CPU does not implement `ne` and hence failed tests.

**Pull request merged**